ark/utils/_bootstrapping.pyx

```diff
@@ -0,0 +1,147 @@
+        _c_permutation(rand_rows, num_choices)
+        _c_permutation(rand_cols, num_choices)
+        memset(rand_cols_flags, 0, num_choices * sizeof(UINT8_t))
+        _init_flag_table(rand_cols_flags, rand_cols, m2n)
+        for m1_label in rand_rows[:m1n]:
+            flat_start = row_indicies[m1_label]
+            flat_end = row_indicies[m1_label + 1]
+            for m2_idx in range(flat_start, flat_end):
+                m2_label = cols_in_row_flat[m2_idx]
+                accum += rand_cols_flags[m2_label]
+        close_num_rand_view[r] = accum
+
+
+@boundscheck(False) # Deactivate bounds checking
+@wraparound(False)  # Deactivate negative indexing
+@cdivision(True)    # Ignore modulo/divide by zero warning
+cdef _compute_close_num_rand(DTYPE_t[:, :] dist_mat_bin, DTYPE_t[:] cols_in_row_flat,
+                             MAXINDEX_t[:] row_indicies, DTYPE_t[:] marker_nums,
+                             int bootstrap_num):
+
+    cdef Py_ssize_t num_markers = marker_nums.shape[0]
+    cdef Py_ssize_t num_choices = dist_mat_bin.shape[0]
+
+    close_num_rand = np.zeros((num_markers, num_markers, bootstrap_num), dtype=np.uint16)
+    cdef DTYPE_t[:, :, :] close_num_rand_view = close_num_rand
+    
+    # allocate marker_label randomization containers
+    cdef Py_ssize_t* rand_rows = <Py_ssize_t*> PyMem_Malloc(num_choices * sizeof(Py_ssize_t))
+    cdef Py_ssize_t* rand_cols = <Py_ssize_t*> PyMem_Malloc(num_choices * sizeof(Py_ssize_t))
+    cdef UINT8_t* rand_cols_flags = <UINT8_t*> PyMem_Malloc(num_choices * sizeof(UINT8_t))
+    if not rand_rows or not rand_cols or not rand_cols_flags:
+        raise MemoryError()
+
+    # allocate 'm2n < avg_rowsize' memory
+    cdef UINT8_t* m2n_small = <UINT8_t*> PyMem_Malloc(num_markers * sizeof(UINT8_t))
+
+    # initialize marker_label containers and m2n_small
+    cdef int mn
+    cdef DTYPE_t avg_rowsize
+    for mn in range(num_choices):
+        rand_rows[mn] = mn
+        rand_cols[mn] = mn
+        avg_rowsize = row_indicies[mn + 1] - row_indicies[mn]
+
+    avg_rowsize /= num_choices
+    for mn in range(num_markers):
+        m2n_small[mn] = (marker_nums[mn] < avg_rowsize)
+
+    cdef int j, k, m1n, m2n
+
+    for j in range(num_markers):
+        m1n = marker_nums[j]
+        for k in range(j, num_markers):
+            m2n = marker_nums[k]
+            if m2n_small[k]:
+                _list_accum(close_num_rand_view[j, k, :], dist_mat_bin, rand_rows, rand_cols,
+                            num_choices, m1n, m2n, bootstrap_num)
+            else:
+                _dict_accum(close_num_rand_view[j, k, :], cols_in_row_flat, row_indicies,
+                            rand_rows, rand_cols, rand_cols_flags, num_choices, m1n, m2n,
+                            bootstrap_num)
+            close_num_rand_view[k, j, :] = close_num_rand_view[j, k, :]
+
+    PyMem_Free(rand_rows)
+    PyMem_Free(rand_cols)
+    PyMem_Free(rand_cols_flags)
+    PyMem_Free(m2n_small)
+
+    return close_num_rand
```

Comment:
Along with Noah's comment above, it would be helpful to add:

- Documentation describing the arguments and return type
- What `m2n_small` is and how it determines whether to use `_list_accum` or `_dict_accum`
- Etc. for anything that might cause confusion